lib/std/os/windows.zig

```diff
@@ -135,32 +136,43 @@ pub fn OpenFileW(
         .SecurityQualityOfService = null,
     };
     var io: IO_STATUS_BLOCK = undefined;
-    const rc = ntdll.NtCreateFile(
-        &result,
-        access_mask,
-        &attr,
-        &io,
-        null,
-        FILE_ATTRIBUTE_NORMAL,
-        FILE_SHARE_WRITE | FILE_SHARE_READ | FILE_SHARE_DELETE,
-        creation,
-        FILE_NON_DIRECTORY_FILE | FILE_SYNCHRONOUS_IO_NONALERT,
-        null,
-        0,
-    );
-    switch (rc) {
-        .SUCCESS => return result,
-        .OBJECT_NAME_INVALID => unreachable,
-        .OBJECT_NAME_NOT_FOUND => return error.FileNotFound,
-        .OBJECT_PATH_NOT_FOUND => return error.FileNotFound,
-        .NO_MEDIA_IN_DEVICE => return error.NoDevice,
-        .INVALID_PARAMETER => unreachable,
-        .SHARING_VIOLATION => return error.SharingViolation,
-        .ACCESS_DENIED => return error.AccessDenied,
-        .PIPE_BUSY => return error.PipeBusy,
-        .OBJECT_PATH_SYNTAX_BAD => unreachable,
-        .OBJECT_NAME_COLLISION => return error.PathAlreadyExists,
-        else => return unexpectedStatus(rc),
+    const share_access = share_access_opt orelse FILE_SHARE_WRITE | FILE_SHARE_READ | FILE_SHARE_DELETE;
+
+    var delay: usize = 1;
+    while (true) {
+        const rc = ntdll.NtCreateFile(
+            &result,
+            access_mask,
+            &attr,
+            &io,
+            null,
+            FILE_ATTRIBUTE_NORMAL,
+            share_access,
+            creation,
+            FILE_NON_DIRECTORY_FILE | FILE_SYNCHRONOUS_IO_NONALERT,
+            null,
+            0,
+        );
+        switch (rc) {
+            .SUCCESS => return result,
+            .OBJECT_NAME_INVALID => unreachable,
+            .OBJECT_NAME_NOT_FOUND => return error.FileNotFound,
+            .OBJECT_PATH_NOT_FOUND => return error.FileNotFound,
+            .NO_MEDIA_IN_DEVICE => return error.NoDevice,
+            .INVALID_PARAMETER => unreachable,
+            .SHARING_VIOLATION => {
+                std.time.sleep(delay);
+                if (delay < 1 * std.time.ns_per_s) {
+                    delay *= 2;
+                }
+                continue; // TODO: don't loop for async
+            },
```

Comment:
2 things here:

1. What about the use case where the API user wants to get `error.SharingViolation` rather than sleep + retry? Likewise, `flock` supports passing `LOCK_NB` and then it returns `EWOULDBLOCK` if the lock is held. The OpenFlags or CreateFlags struct should probably have another flag there to control whether the code waits for the lock or returns an error if it cannot obtain the lock.

2. I'm OK with this sleep loop for now, for the case of when the user sets the flag to wait for the lock, but I'll file a follow-up issue for this. Unconditional sleeping is a code smell, and I can definitely think of use cases where the API user will need to abort this retry loop. This solves the use case that we have in the compiler, but it's not the final API that the std lib should end up on. But this does not need to block the merge of this PR.

Reply:
I'm okay implementing this, but I didn't implement it initially because the only other code that related to blocking was `async`/`await` stuff. Thinking about it now, though, wanting it to not block and `async` are different things, and shouldn't depend on one another.

As for number 2, yeah, it's pretty ugly. Unfortunately, windows has no support for waiting until the file is available. We could make both linux and windows return immediately instead of blocking, or simply make it a compile error to open a file with a lock and not in non blocking mode on windows.